lib/core/src/Cardano/Wallet/Primitive/AddressDerivation/Icarus.hs

```diff
@@ -151,6 +172,84 @@ generateKeyFromSeed
     -> IcarusKey 'RootK XPrv
 generateKeyFromSeed = unsafeGenerateKeyFromSeed
 
+-- | Hardware Ledger devices generates keys from mnemonic using a different
+-- approach (different from the rest of Cardano).
+generateKeyFromHardwareLedger
+    :: Mnemonic 24
+        -- ^ The actual seed
+    -> Passphrase "encryption"
+        -- ^ Master encryption passphrase
+    -> IcarusKey 'RootK XPrv
+generateKeyFromHardwareLedger mnemonic (Passphrase pwd) = unsafeFromRight $ do
+    let seed = pbkdf2HmacSha512
+            $ T.encodeUtf8
+            $ T.intercalate " "
+            $ mnemonicToText mnemonic
+
+    let cc = hmacSha256 (BS.pack [1] <> seed)
+    let (kL, kR) = first normalizeEd25519 $ hashRepeatedly seed
+    pA <- ed25519ScalarMult kL
+
+    prv <- left show $ xprv $ kL <> kR <> pA <> cc
+    pure $ IcarusKey (xPrvChangePass (mempty :: ByteString) pwd prv)
+  where
+    unsafeFromRight :: (HasCallStack, Show e) => Either e a -> a
+    unsafeFromRight = either (error . show) id
+
+    hashRepeatedly :: ByteString -> (ByteString, ByteString)
+    hashRepeatedly bytes = case BS.splitAt 32 (hmacSha512 bytes) of
+        (kL, kR) | testBit (kL `BS.index` 31) 5 -> hashRepeatedly (kL <> kR)
```

Comment:
Perhaps we could clarify the guard condition by putting it into a `where` clause and calling it `isInvalidKey`.

From SLIP-0010:

> To avoid invalid master keys, the algorithm is retried with the intermediate hash as new seed if the key is invalid.
> ....
> For ed25519, the last step always succeeds since every 256-bit number (even 0) is a valid private key.

Given that, what is this testBit for? It excludes keys which are multiples of 32 ... are these known to be weak keys?

Reply:
To be honest. I have no idea and I couldn't find any satisfactory piece of documentation about this. Only thing I know is that the Ledger implementation seems to do this, for... a mysterious reason.